src/Control/Carrier/NonDet/Church.hs

```diff
@@ -73,14 +72,17 @@ instance Fail.MonadFail m => Fail.MonadFail (NonDetC m) where
   fail s = lift (Fail.fail s)
   {-# INLINE fail #-}
 
+-- | Separate fixpoints are computed for each branch.
+--
+-- >>> run (runNonDetA @[] (take 3 <$> mfix (\ as -> pure (0 : map succ as) <|> pure (0 : map pred as))))
+-- [[0,1,2],[0,-1,-2]]
 instance MonadFix m => MonadFix (NonDetC m) where
   mfix f = NonDetC $ \ fork leaf nil ->
-    mfix (runNonDet
-      (liftA2 Fork)
-      (pure . Leaf)
-      (pure Nil)
-      . f . fromJust . fold (<|>) Just Nothing)
-    >>= fold fork leaf nil
+    mfix (runNonDetA . f . head)
+    >>= runNonDet fork leaf nil . foldr
+      (\ a _ -> pure a <|> mfix (liftAll . fmap tail . runNonDetA . f))
+      empty where
+    liftAll m = NonDetC $ \ fork leaf nil -> m >>= foldr (fork . leaf) nil
```

Comment:
This was pretty tricky to figure out; `ListT`, of all things, saved the day. Essentially what we do is run `f` once per branch it returns, `<|>`ing their results up, to ensure that each branch computes a separate fixpoint.